doc/dev/haskell-idioms.md

```diff
@@ -0,0 +1,182 @@
+# Fancy Haskell idioms
+
+This page lists a few idioms that are used in GREASE and might be a bit
+surprising when you first see them. To understand these idioms, you should be
+comfortable with GADTs, kinds, and monads.
+
+## Types with names that start with `Some`
+
+Haskell currently lacks support for [first-class existential types]. They can be
+emulated with [GADT]s like so:
+```haskell
+{-# LANGUAGE GADTs #-}
+
+data Some f where
+  Some :: f x -> Some f  -- the type does not mention `x`
+```
+(`Some` is canonically available as [`Data.Parameterized.Some.Some`].) As
+defined above, `Some` has kind `forall k. (k -> Type) -> Type`. What if you want
+to existentially quantify over something of a different kind, e.g., you want to
+close over the first parameter of a datatype with two parameters? Then, you have
+to make your own GADT, and such types are conventionally named `Some*`.
+
+[first-class existential types]: https://github.com/ghc-proposals/ghc-proposals/pull/473
+
+## Faux `let`-bindings in type signatures
+
+Sometimes, a long type like `AVeryLongTypeName a (AnotherLongName b)` appears
+several times in one signature:
+```haskell
+foo ::
+  ( Eq (AVeryLongTypeName a (AnotherLongName b))
+  , Hashable (AVeryLongTypeName a (AnotherLongName b))
+  , Show (AVeryLongTypeName a (AnotherLongName b))
+  ) =>
+  AVeryLongTypeName a (AnotherLongName b) ->
+  AVeryLongTypeName a (AnotherLongName b) ->
+  ()
+foo _x _y = ()
+```
+One could define a type synonym
+```haskell
+type ShorterName a b = AVeryLongTypeName a (AnotherLongName b)
+```
+However, this forces readers of the code to know about (and remember) the
+synonym. Alternatively, one can use equality constraints to form a "faux
+`let`-binding" like so:
+```haskell
+foo ::
+  ( t ~ AVeryLongTypeName a (AnotherLongName b)
+  , Eq t
+  , Hashable t
+  , Show t
+  ) =>
+  t ->
+  t ->
+  ()
+foo _x _y = ()
+```
+
+## Pattern matching on `pure` values in `do` blocks
+
+[GADT] constructors can existentially quantify types.
+```haskell
+{-# LANGUAGE GADTs #-}
+
+data Some f where
+  Some :: f x -> Some f  -- the type does not mention `x`
+```
+(`Some` is canonically available as [`Data.Parameterized.Some.Some`].) To use
+a `Some`, you generally pattern match on it. This brings the captured variable
+into scope (though just as an unconstrained variable, so any code you write
+after the pattern match has to be polymorphic over it).
+```haskell
+wasItJust :: Some Maybe -> String
+wasItJust someMaybe =
+  case someMaybe of
+    -- Can use _x here, but have to be completely polymorphic over its type. Not
+    -- really much you *can* do with a value of an unknown type.
+    Some (Just _x) -> "it was Just"
+    Some Nothing -> "it was Nothing"
```

Comment:
We might consider using an example where we _do_ use a value with an unknown type, e.g.,

```hs
someLength :: Some [] -> Int
someLength someList =
  case someList of
    -- Can use _x here, but you have to be completely polymorphic over its type.
    -- We use the `length` function, which works for list values of type `[x]` for all
    -- types `x`.
    Some l -> length l
```

Reply:
Good idea, thanks!